src/runtime/execution.rs

```diff
@@ -477,14 +486,21 @@ impl ExecutionState {
             _ => {}
         }
 
+        let mut blocked_attached = false;
         let runnable = self
             .tasks
             .iter()
+            .inspect(|t| blocked_attached = blocked_attached || (t.blocked() && !t.detached))
             .filter(|t| t.runnable())
             .map(|t| t.id)
             .collect::<SmallVec<[_; DEFAULT_INLINE_TASKS]>>();
 
-        if runnable.is_empty() {
+        // We should finish execution when either
+        // (1) There are no runnable tasks, or
+        // (2) All runnable tasks have been detached AND there are no blocked attached tasks
+        // If there are some blocked attached tasks and all runnable tasks are detached,
+        // we must run some detached task so that blocked attached tasks may become unblocked.
+        if runnable.is_empty() || (!blocked_attached && runnable.iter().all(|id| self.get(*id).detached)) {
```

Comment:
Now that I think about it some more, does this logic go away once we implement detached threads? I'm thinking that once that happens, the logic for "finished execution" is just "main thread is finished", because every other thread has either been joined on by the main thread (perhaps transitively), or has been detached.

Reply:
I see, the (2) case would just boil down to the main thread being finished I think. Maybe we keep this here for now and change the logic in a separate "detached threads" PR?

Reply:
Yeah, we can't make it work that way without doing the detached threads thing, so it makes sense to defer it to later.